### ISSUE TITLE
Choppy playback/framedrops with videos with fps close to the display refresh rate (the same like issue #3739 for osx)

### ISSUE THREAD

**Issue by jcapik** [mpv-player/mpv#10588]
mpv version = 0.32.0 (0.32.0-1ubuntu1)
Linux Distribution and Version = Linux Mint 20.3 (=Ubuntu 20.04 Focal)
Source of the mpv binary = oficial Ubuntu repo
which version of mpv introduced the problem = unknown
Window Manager and version = xfce 4.16
GPU driver and version = nvidia-driver-390 (390.154-0ubuntu5.20.04.2)

Reproduction steps:
set the display FPS to 23.976Hz and play 24Hz video with the vdpau output driver

Expected behavior:
smooth playback

Actual behavior:
choppy playback (display-adrop helps a bit, but still there)

Log file:
https://0x0.st/opT0.txt

Note: Works well in VLC

**Comment by Dudemanguy** [comment]
0.32 is way too old but regardless don't use the vdpau vo. What you probably want is ``--hwdec=vpdau``.

**Comment by jcapik** [comment]
The issue is present even when vo is set to opengl or XV ...

**Comment by Dudemanguy** [comment]
We don't support outdated mpv versions.

**Comment by jcapik** [comment]
Well, I can re-test with the latest if that's required ... are we still people able to talk before acting?

**Comment by Dudemanguy** [comment]
If you can reproduce with the master branch on `vo=gpu` feel free to reopen.

**Comment by jcapik** [comment]
Maybe such info could be included in the template ?

**Comment by christoph-heinrich** [comment]
It does include
> If you're not using git master or the latest release, update.
> Releases are listed here: https://github.com/mpv-player/mpv/releases

**Comment by jcapik** [comment]
Ok, so you essentially only allow issue reports from mpv users who know, how to build mpv from sources on their distribution of choice, right? Regardless of the fact the build instructions are incomplete and the waf configure says x11 is missing even when the libx11-dev is installed :] ... Do you really expect majority of the users want to undergo such martyrdom just because they want to report an issue? To me it looked a bit like "we're not interested in your reports ... leave us alone" :)

**Comment by christoph-heinrich** [comment]
It seems you missed the
> or the latest release

part of that quote.

**Comment by jcapik** [comment]
[Dudemanguy] explicitly stated master branch and vo=gpu ... 

Yesterday I upgraded the whole distro just because of this report and tested with 0.34.1 and the problem is still present

**Comment by jcapik** [comment]
The log from 0.34.1 is here ... https://0x0.st/opuH.txt 

So, may I please ask you to re-open this issue?

**Comment by Dudemanguy** [comment]
Can you post a full log file (--log-file) using vo=gpu?

**Comment by jcapik** [comment]
I've finally managed to find out the missing header and the mpv build was successful ... the issue is NOT present in master .... that means I'll have to stick with the built binary and convince the ubuntu maintainer to upgrade or at least backport the relevant patches ...

**Comment by jcapik** [comment]
> Can you post a full log file (--log-file) using vo=gpu?

with regard to the above, it is probably no longer relevant, but ...

https://0x0.st/opu1.txt

**Comment by jcapik** [comment]
Sorry, I have to take my above statement back ... it seems the issue is related to multiple display ...

**Comment by jcapik** [comment]
The following is my xrandr output ...

$ xrandr
Screen 0: minimum 8 x 8, current 3840 x 1200, maximum 16384 x 16384
DVI-I-0 disconnected (normal left inverted right x axis y axis)
DVI-I-1 connected primary 1920x1200+0+0 (normal left inverted right x axis y axis) 489mm x 298mm
   1920x1200     59.95*+
   1920x1080     59.94    50.00  
   1680x1050     59.95  
   1600x1200     60.00  
   1600x900      60.00  
   1280x1024     60.02  
   1280x800      59.81  
   1280x720      60.00    59.94    50.00  
   1024x768      60.00  
   800x600       60.32  
   720x576       50.00  
   720x480       59.94  
   640x480       59.94  
DP-0 disconnected (normal left inverted right x axis y axis)
HDMI-0 connected 1920x1080+1920+0 (normal left inverted right x axis y axis) 1600mm x 900mm
   1920x1080     60.00 +  59.94    50.00    29.97    25.00    23.98*   60.05    60.00    50.04  
   1360x768      60.02  
   1280x1024     60.02  
   1280x720      60.00    59.94    50.00  
   1024x768      60.00  
   800x600       60.32  
   720x576       50.00  
   720x480       59.94  
   640x480       59.95    59.94    59.93  
DP-1 disconnected (normal left inverted right x axis y axis)

**Comment by Dudemanguy** [comment]
mpv appears to use the correct display FPS according to your log. Since you said you were on NVIDIA, try using the `__GL_SYNC_DISPLAY_DEVICE` variable to sync to the correct monitor.

**Comment by jcapik** [comment]
> mpv appears to use the correct display FPS according to your log. Since you said you were on NVIDIA, try using the `__GL_SYNC_DISPLAY_DEVICE` variable to sync to the correct monitor.

That variable I set a long time ago and of course it contains the correct value "DFP-2" pointing at the right display. Thanks to that I get no tearing at least with some of the output drivers.

**Comment by jcapik** [comment]
Just to clarify what I mean by "choppy" ...  I'm not talk about tearing ... I'm talking about something that looks like dropping frames

**Comment by jcapik** [comment]
When I use --video-sync=display-resample or --video-sync=display-adrop, it behaves better ... that means the frame drops are kinda spread ... not absolutely equally/evenly, but it's much better than scenario without that option when it accumulates and then  I get a very visible sequence of framedrops concentrated at the same time

**Comment by Dudemanguy** [comment]
>That variable I set a long time ago and of course it contains the correct value "DFP-2" pointing at the right display.

It's not `HDMI-0` or something like that?

**Comment by jcapik** [comment]
> > That variable I set a long time ago and of course it contains the correct value "DFP-2" pointing at the right display.
> 
> It's not `HDMI-0` or something like that?

Nope ... it eats Xorg device names ...

[ 21308.072] (--) NVIDIA(GPU-0): Eizo EV2360 (DFP-0): connected
[ 21308.072] (--) NVIDIA(GPU-0): Eizo EV2360 (DFP-0): Internal TMDS
[ 21308.072] (--) NVIDIA(GPU-0): Eizo EV2360 (DFP-0): 225.0 MHz maximum pixel clock
[ 21308.072] (--) NVIDIA(GPU-0):
[ 21308.072] (--) NVIDIA(GPU-0): DFP-1: disconnected
[ 21308.072] (--) NVIDIA(GPU-0): DFP-1: Internal TMDS
[ 21308.072] (--) NVIDIA(GPU-0): DFP-1: 165.0 MHz maximum pixel clock
[ 21308.072] (--) NVIDIA(GPU-0):
[ 21308.104] (--) NVIDIA(GPU-0): LG Electronics LG TV (DFP-2): connected
[ 21308.104] (--) NVIDIA(GPU-0): LG Electronics LG TV (DFP-2): Internal TMDS
[ 21308.104] (--) NVIDIA(GPU-0): LG Electronics LG TV (DFP-2): 225.0 MHz maximum pixel clock

**Comment by Dudemanguy** [comment]
Oh well I guess NVIDIA is weird. Anyways, I'll reopen this I suppose.

**Comment by jcapik** [comment]
> Oh well I guess NVIDIA is weird. Anyways, I'll reopen this I suppose.

Yeah, I agree ... I was quite convinced to never buy any Nvidia card again till the point when I got an info from a friend of mine they wanted to open their closed driver ....finally

**Comment by ZaquL** [comment]
23.977374 FPS for display sync. 

[ 0.088][v][vd] Container reported FPS: 24.000002

You cannot do that. Not only your display is not ideal 24/1.001, but your container is not ideal 24 fps. But anyway, 24.000 fps should be played on 24.000 display, not 24/1.001 display.

**Comment by kevmitch** [comment]
> There is no simple way to convert between the two.

Yes there is. That's what [`--video-sync=display-*`](https://mpv.io/manual/master/#options-video-sync) is supposed to do.

**Comment by ZaquL** [comment]
In fact this is below threshold and so any interpolation turns off (it should print that). I can say for sure if you will give me the file or part of it.

**Comment by christoph-heinrich** [comment]
`--video-sync=display-*` does not perform any interpolation

**Comment by jcapik** [comment]
> `--video-sync=display-*` does not perform any interpolation

--video-sync=display-resample does perform a sound interpolation .... if I understand that correctly, it should slow the movie down so that the video fps matches the screen fps exactly and the sound is resampled to be in sync with the slower fps.

**Comment by jcapik** [comment]
One more note ... I've tested the video in mplayer and it works correctly too. It seems mpv is the only player that exhibits this issue

**Comment by jcapik** [comment]
> > In fact this is below threshold and so any interpolation turns off (it should print that). I can say for sure if you will give me the file or part of it.
> 
> Is it? I was reading about that and the threshold should be sufficient for the 23.976<->24 and 29.97<->30 difference. In fact, I've tried to increase the threshold to a value allowing 23.976<->25 speedup/slowdown ...

I hope you mean the --video-sync-max-video-change=10 threshold.

I've tested to play the video with 25Hz and the problem was still there. The sound was apparently higher, but still acceptable

**Comment by jcapik** [comment]
One more thing. I have one more video with 24.000002Hz and it doesn't exhibit such problems ... Could that be a problem caused by wrong/inaccurate values in the header? Both videos were encoded by the same person and with the same software.

**Comment by christoph-heinrich** [comment]
> --video-sync=display-resample does perform a sound interpolation

Not sure if resampling counts as interpolation, but regardless I thought you meant the behavior of the ``--interpolation`` option.

> if I understand that correctly, it should slow the movie down so that the video fps matches the screen fps exactly and the sound is resampled to be in sync with the slower fps and that requires interpolation

It may speed up or slow down playback, but other then that you understood it correctly.
Technically it tries to line up new frames from the video with the refreshes of the monitor, which is not quite the same (after all it's not like would play 24fps video at 2.5 times speed to reach 60fps, even if it had enough leeway to do so), but it's close enough.

**Comment by jcapik** [comment]
> > --video-sync=display-resample does perform a sound interpolation
> 
> Not sure if resampling counts as interpolation,

I'm pretty positive about that :] If you change the sampling rate, then you need to calculate the new values of the samples since they (mostly) point somewhere between the original samples. The cheapest sound interpolation is linear, but it produces a distortion. Anyways it's still better than truncating or rounding the sample index and taking such values directly without interpolating them.

> but regardless I thought you meant the behavior of the `--interpolation` option.

Nope, video interpolation is something way more difficult and it was not enabled during my tests

**Comment by jcapik** [comment]
So, what to do next? I could either do some troubleshooting according to your hints or apply diagnostic patches on top of the master branch ... or anything else

**Comment by desbma** [comment]
I think I am hit by this issue too.

I'm on Arch Linux with mpv 0.34.1. I use mpv a lot and only noticed this recently so this is definitely a regression.

The playback is very choppy, and audio/video are desynchronized, then it becomes normal after a few seconds. It seems to only happen after I seek back in the video.

I tried playing with `vo` and `video-sync` options, but that did not fixed it.

**Comment by ZaquL** [comment]
Is --vo gpu-next affected too?

>I have one more video with 24.000002Hz and it doesn't exhibit such problems 

That would mean we really need a sample.

**Comment by jcapik** [comment]
> I think I am hit by this issue too.
> 
> I'm on Arch Linux with mpv 0.34.1. I use mpv a lot and only noticed this recently so this is definitely a regression.
> 
> The playback is very choppy, and audio/video are desynchronized, then it becomes normal after a few seconds. It seems to only happen after I seek back in the video.
> 
> I tried playing with `vo` and `video-sync` options, but that did not fixed it.

That looks like a different issue than mine ...

**Comment by jcapik** [comment]
> Is --vo gpu-next affected too?
> 
> > I have one more video with 24.000002Hz and it doesn't exhibit such problems
> 
> That would mean we really need a sample.

**Comment by jcapik** [comment]
> Is --vo gpu-next affected too?
> 
> > I have one more video with 24.000002Hz and it doesn't exhibit such problems
> 
> That would mean we really need a sample.

Not suitable to share publicly. PM ?

**Comment by ZaquL** [comment]
First please show --vo gpu-next log that shows interpolation turned off.

E.g. "Detected fps ratio 0.001 below threshold 0.0100, disabling interpolation"

And also paste the last lines of ffmpeg -i in -vf vfrdet -an -f null -

here, maybe it is VFR. In that case interpolation is so much harder unless you use VRR (linux and windows and now MacOS support it).

**Comment by jcapik** [comment]
Not a single line about the interpolation ... https://0x0.st/opDj.txt

ffmpeg output:
[Parsed_vfrdet_0 @ 0x55979c152ac0] VFR:0.665722 (470/236) min: 41 max: 42 avg: 41

**Comment by jcapik** [comment]
I tried ffprobe to get the average framerate and ...

$ ffprobe -v quiet -print_format json -show_streams 1080p.mkv | grep avg_frame_rate
            "avg_frame_rate": "24/1",

If I understand that correctly the above means the video is in fact CFR, even when it says it's VFR, right?

**Comment by jcapik** [comment]
Any further steps needed here? Or is it all from my side?
